sweep/tx_input_set.go

```diff
@@ -543,3 +551,217 @@ func createWalletTxInput(utxo *lnwallet.Utxo) (input.Input, error) {
 		&utxo.OutPoint, witnessType, signDesc, heightHint,
 	), nil
 }
+
+// BudgetInputSet implements the interface `InputSet`. It takes a list of
+// pending inputs which share the same deadline height and groups them into a
+// set conditionally based on their economical values.
+type BudgetInputSet struct {
+	// inputs is the set of inputs that have been added to the set after
+	// considering their economical contribution.
+	inputs []*pendingInput
+
+	// budget is the total budget that can be spent as fees.
+	budget btcutil.Amount
+
+	// deadlineHeight is the height which the inputs in this set must be
+	// confirmed by.
+	deadlineHeight fn.Option[int32]
+}
+
+// Compile-time constraint to ensure budgetInputSet implements InputSet.
+var _ InputSet = (*BudgetInputSet)(nil)
+
+// NewBudgetInputSet creates a new BudgetInputSet.
+//
+// NOTE: panic if supplied inputs slice is empty.
+func NewBudgetInputSet(inputs []*pendingInput) *BudgetInputSet {
+	// TODO(yy): all the inputs share the same deadline height, which means
+	// there exists an opportunity to refactor the deadline height to be
+	// tracked on the set-level, not per input. This would allow us to
+	// avoid the overhead of tracking the same height for each input in the
+	// set.
+	deadlineHeight := inputs[0].params.DeadlineHeight
+	bi := &BudgetInputSet{
+		deadlineHeight: deadlineHeight,
+		inputs:         make([]*pendingInput, 0, len(inputs)),
+	}
+
+	for _, input := range inputs {
+		bi.addInput(input)
+	}
+
+	log.Tracef("Created %v", bi.String())
+
+	return bi
+}
+
+// String returns a human-readable description of the input set.
+func (b *BudgetInputSet) String() string {
+	deadlineDesc := "none"
+	b.deadlineHeight.WhenSome(func(h int32) {
+		deadlineDesc = fmt.Sprintf("%d", h)
+	})
+
+	inputsDesc := ""
+	for _, input := range b.inputs {
+		inputsDesc += fmt.Sprintf("\n%v", input)
+	}
+
+	return fmt.Sprintf("BudgetInputSet(budget=%v, deadline=%v, "+
+		"inputs=[%v])", b.budget, deadlineDesc, inputsDesc)
+}
+
+// addInput adds an input to the input set.
+func (b *BudgetInputSet) addInput(input *pendingInput) {
+	b.inputs = append(b.inputs, input)
+	b.budget += input.params.Budget
+}
+
+// NeedWalletInput returns true if the input set needs more wallet inputs.
+func (b *BudgetInputSet) NeedWalletInput() bool {
+	var (
+		// budgetNeeded is the amount that needs to be covered from
+		// other inputs.
+		budgetNeeded btcutil.Amount
+
+		// budgetBorrowable is the amount that can be borrowed from
+		// other inputs.
+		budgetBorrowable btcutil.Amount
+	)
+
+	for _, inp := range b.inputs {
+		// If this input has a required output, we can assume it's a
+		// second-level htlc txns input. Although this input must have
+		// a value that can cover its budget, it cannot be used to pay
+		// fees. Instead, we need to borrow budget from other inputs to
+		// make the sweep happen. Once swept, the input value will be
+		// credited to the wallet.
```

Comment:
We can do that, actually initially thought about doing it, then felt it was too specific - I think the sweeper only wanna look at certain properties of a given input, and SS-HTLC is abstracted as `RequiredTxOut`. Tho I think it's not that much of a good abstraction and a bit overdesign since only SS-HTLCs use this property.